core/src/main/java/nu/yona/server/subscriptions/service/UserAdditionService.java

```diff
@@ -0,0 +1,275 @@
+
+		User userEntity = createUserEntity(user, UserSignUp.FREE);
+		addDevicesToEntity(user, userEntity);
+		Optional<ConfirmationCode> confirmationCode = createConfirmationCodeIfNeeded(overwriteUserConfirmationCode, userEntity);
+		userEntity = userRepository.save(userEntity);
+		UserDto userDto = userRetrievalService.createUserDtoWithPrivateData(userEntity);
+		if (confirmationCode.isPresent())
+		{
+			userUpdateService.sendConfirmationCodeTextMessage(userEntity.getMobileNumber(), confirmationCode.get(),
+					SmsTemplate.ADD_USER_NUMBER_CONFIRMATION);
+		}
+
+		logger.info("Added new user with mobile number '{}' and ID '{}'", userDto.getMobileNumber(), userDto.getId());
+		return userDto;
+	}
+
+	private User createUserEntity(UserDto user, UserSignUp signUp)
+	{
+		byte[] initializationVector = CryptoSession.getCurrent().generateInitializationVector();
+		MessageSource anonymousMessageSource = MessageSource.getRepository().save(MessageSource.createInstance());
+		MessageSource namedMessageSource = MessageSource.getRepository().save(MessageSource.createInstance());
+		Set<Goal> goals = buildGoalsSet(user, signUp);
+		UserAnonymized userAnonymized = UserAnonymized.createInstance(anonymousMessageSource.getDestination(), goals);
+		UserAnonymized.getRepository().save(userAnonymized);
+		UserPrivate userPrivate = UserPrivate.createInstance(user.getOwnPrivateData().getFirstName(),
+				user.getOwnPrivateData().getLastName(), user.getOwnPrivateData().getNickname(), userAnonymized.getId(),
+				anonymousMessageSource.getId(), namedMessageSource);
+		User userEntity = new User(UUID.randomUUID(), initializationVector, user.getMobileNumber(), userPrivate,
+				namedMessageSource.getDestination());
+		addMandatoryGoals(userEntity);
+		return userEntity;
+	}
+
+	private void addDevicesToEntity(UserDto userDto, User userEntity)
+	{
+		userRepository.saveAndFlush(userEntity); // To prevent sequence issues when saving the device
+		userDto.getOwnPrivateData().getDevices().orElse(Collections.emptySet()).stream().map(d -> (UserDeviceDto) d)
+				.forEach(d -> deviceService.addDeviceToUser(userEntity, d));
+	}
+
+	private Set<Goal> buildGoalsSet(UserDto user, UserSignUp signUp)
+	{
+		Set<Goal> goals;
+		if (signUp == UserSignUp.INVITED)
+		{
+			goals = Collections.emptySet();
+		}
+		else
+		{
+			goals = user.getOwnPrivateData().getGoals().orElse(Collections.emptySet()).stream().map(GoalDto::createGoalEntity)
+					.collect(Collectors.toSet());
+		}
+		return goals;
+	}
+
+	private Optional<ConfirmationCode> createConfirmationCodeIfNeeded(Optional<String> overwriteUserConfirmationCode,
+			User userEntity)
+	{
+		if (overwriteUserConfirmationCode.isPresent())
+		{
+			// no need to confirm again
+			userEntity.markMobileNumberConfirmed();
+			return Optional.empty();
+		}
+
+		Optional<ConfirmationCode> confirmationCode = Optional.of(userUpdateService.createConfirmationCode());
+		userEntity.setMobileNumberConfirmationCode(confirmationCode.get());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Call "confirmationCode.isPresent()" before accessing the value. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AS3655)